src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,49 +1,113 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_LENGTH = 10;
+    private static final int LIST_MAX_SIZE = Integer.MAX_VALUE - 8;
+    private int size;
+    private T[] array;
+
+    public ArrayList() {
+        this.array = (T[]) new Object[DEFAULT_LENGTH];
+        size = 0;
+    }
 
     @Override
     public void add(T value) {
-
+        if (size == array.length) {
+            expandArray();
+        }
+        array[size] = value;
+        size += 1;
     }
 
     @Override
     public void add(T value, int index) {
-
+        size += 1;
+        if (array.length <= size) {
+            expandArray();
+        }
+        indexValidation(index);
+        T[] newArray = Arrays.copyOf(array, array.length);
+        newArray[index] = value;
+        System.arraycopy(array, index, newArray, index + 1, size - index);
+        array = newArray;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexValidation(index);
+        return array[index];
     }
 
+    /*
+    if value wasn't found in list then no actions provided
+     */
     @Override
     public void set(T value, int index) {
-
+        indexValidation(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexValidation(index);
+        T removed = array[index];
+        System.arraycopy(array, index + 1, array, index, size - index);
+        size -= 1;
+        return removed;
     }
 
     @Override
-    public T remove(T t) {
-        return null;
+    public T remove(T element) {
+        int index = getIndex(element);
+        if (index == -1) {
+            throw new NoSuchElementException("The "
+                    + element.toString() + " wasn't found in the list!");
+        }
+        return index == -1 ? null : remove(index);
```

Comment:
what will happen if index == -1 at 74 line? is there a chance that at this line (77) we can get index = -1?

Reply:
true, this condition is redundant